node-js-basics-101/node_modules/ws/lib/WebSocket.js

```diff
@@ -0,0 +1,712 @@
+      this.close(1002, '');
+      this.emit('error', error);
+    };
+
+    this.readyState = WebSocket.OPEN;
+    this.emit('open');
+  }
+
+  /**
+   * Clean up and release internal resources.
+   *
+   * @param {(Boolean|Error)} Indicates whether or not an error occurred
+   * @private
+   */
+  finalize (error) {
+    if (this._finalizeCalled) return;
+
+    this.readyState = WebSocket.CLOSING;
+    this._finalizeCalled = true;
+
+    clearTimeout(this._closeTimer);
+    this._closeTimer = null;
+
+    //
+    // If the connection was closed abnormally (with an error), or if the close
+    // control frame was malformed or not received then the close code must be
+    // 1006.
+    //
+    if (error) this._closeCode = 1006;
+
+    if (this._socket) {
+      this._ultron.destroy();
+      this._socket.on('error', function onerror () {
+        this.destroy();
+      });
+
+      if (!error) this._socket.end();
+      else this._socket.destroy();
+
+      this._socket = null;
+      this._ultron = null;
+    }
+
+    if (this._sender) {
+      this._sender = this._sender.onerror = null;
+    }
+
+    if (this._receiver) {
+      this._receiver.cleanup(() => this.emitClose());
+      this._receiver = null;
+    } else {
+      this.emitClose();
+    }
+  }
+
+  /**
+   * Emit the `close` event.
+   *
+   * @private
+   */
+  emitClose () {
+    this.readyState = WebSocket.CLOSED;
+    this.emit('close', this._closeCode || 1006, this._closeMessage || '');
+
+    if (this.extensions[PerMessageDeflate.extensionName]) {
+      this.extensions[PerMessageDeflate.extensionName].cleanup();
+    }
+
+    this.extensions = null;
+
+    this.removeAllListeners();
+    this.on('error', constants.NOOP); // Catch all errors after this.
+  }
+
+  /**
+   * Pause the socket stream.
+   *
+   * @public
+   */
+  pause () {
+    if (this.readyState !== WebSocket.OPEN) throw new Error('not opened');
+
+    this._socket.pause();
+  }
+
+  /**
+   * Resume the socket stream
+   *
+   * @public
+   */
+  resume () {
+    if (this.readyState !== WebSocket.OPEN) throw new Error('not opened');
+
+    this._socket.resume();
+  }
+
+  /**
+   * Start a closing handshake.
+   *
+   * @param {Number} code Status code explaining why the connection is closing
+   * @param {String} data A string explaining why the connection is closing
+   * @public
+   */
+  close (code, data) {
+    if (this.readyState === WebSocket.CLOSED) return;
+    if (this.readyState === WebSocket.CONNECTING) {
+      if (this._req && !this._req.aborted) {
+        this._req.abort();
+        this.emit('error', new Error('closed before the connection is established'));
+        this.finalize(true);
+      }
+      return;
+    }
+
+    if (this.readyState === WebSocket.CLOSING) {
+      if (this._closeCode && this._socket) this._socket.end();
+      return;
+    }
+
+    this.readyState = WebSocket.CLOSING;
+    this._sender.close(code, data, !this._isServer, (err) => {
```

Comment:
__[jshint]__
* 'arrow function syntax (=>)' is only available in ES6 (use 'esversion: 6').

<!-- Comment by [SideCI](https://sideci.com) -->